### PR TITLE
Enhance Color types

### PR DESCRIPTION
**Pull request by TheSonOfThomp**
## ✍️ Proposed changes

`lib`: 
 - Adds `RecursiveRecord` type

`tokens`:
 - Improves semantics of color token segment type names (e.g. `Type` -> `Property`, `State` -> `InteractionState`).
 - Adds TSDoc to color types
 - Deprecates `tokens.Mode` in favor of `lib.Theme`

### CODE REVIEW

File: packages/lib/src/types/RecursiveRecord.d.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+ *     >>
+ *   >>
+ * >>
+ * ```
+ * */
+export type RecursiveRecord<
+  Keys extends Array<any>,
+  Strict extends boolean = true,
+> = Keys extends [
+  infer Key, // the current union of keys
```
Comment (shaneeza) [inline packages/lib/src/types/RecursiveRecord.d.ts:28]: Can you ELI5 what's happening on lines 27-30?

Comment (TheSonOfThomp) [reply packages/lib/src/types/RecursiveRecord.d.ts:28]: > If the type `Keys` is an array, then de-structure that type. The first index in the array is an inferred `Key` type, and then a spread the remaining indexes into an inferred `Rest` type.

Comment (TheSonOfThomp) [reply packages/lib/src/types/RecursiveRecord.d.ts:28]: Technically `Keys` is not an array, but a [Variadic Tuple](https://typescriptlang.org/docs/handbook/release-notes/typescript-4-0.html#variadic-tuple-types)

Comment (shaneeza) [reply packages/lib/src/types/RecursiveRecord.d.ts:28]: That was very helpful; thanks for the breakdown!
